### PR TITLE
feat: terraform contribution guide

### PR DESCRIPTION
**Pull request by segraef**
# Overview/Summary

## This PR changes

1. Terraform contribution guide

If you want to view the compiled version of the current page clone/check [this](https://github.com/segraef/Azure-Verified-Modules/tree/feat/32507) and run `hugo serve`.

## As part of this Pull Request I have

- [x] Read the Contribution Guide and ensured this PR is compliant with the guide
- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/Azure-Verified-Modules/pulls)
- [ ] Associated it with relevant [GitHub Issues](https://github.com/Azure/Azure-Verified-Modules/issues) or ADO Work Items (Internal Only)
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/Azure-Verified-Modules/)
- [x] Ensured PR tests are passing
- [x] Updated relevant and associated documentation (e.g. Contribution Guide, Docs etc.)

### CODE REVIEW

**Review by jtracey93 — DISMISSED** [review]
LGTM @matt-FFFFFF & @mbilalamjad thoughts?

File: docs/content/contributing/terraform/composition.md
Comment (matt-FFFFFF) [inline docs/content/contributing/terraform/composition.md:155]: Do we need to repeat the specs in here? Seems like double effort.

File: docs/content/contributing/terraform/composition.md
Comment (matt-FFFFFF) [inline docs/content/contributing/terraform/composition.md:254]: This is duplicated above?

File: docs/content/contributing/terraform/prerequisites.md
Comment (matt-FFFFFF) [inline docs/content/contributing/terraform/prerequisites.md:62]: I think this should be removed

File: docs/content/contributing/terraform/prerequisites.md
Comment (matt-FFFFFF) [inline docs/content/contributing/terraform/prerequisites.md:71]: Does not need to be manually installed, will be part of container

File: docs/content/contributing/terraform/prerequisites.md
Comment (matt-FFFFFF) [inline docs/content/contributing/terraform/prerequisites.md:74]: This doesn't work well with WSL (or didn't) can we check?

File: docs/content/contributing/terraform/terraform-contribution-flow/_index.md
Comment (matt-FFFFFF) [inline docs/content/contributing/terraform/terraform-contribution-flow/_index.md:105]: Not if they are a member of the contributors group

File: docs/content/contributing/terraform/terraform-contribution-flow/_index.md
Comment (matt-FFFFFF) [inline docs/content/contributing/terraform/terraform-contribution-flow/_index.md:147]: Need to revise this based on new runners - ping me so we can work it out.

File: docs/content/contributing/terraform/terraform-contribution-flow/_index.md
Comment (matt-FFFFFF) [inline docs/content/contributing/terraform/terraform-contribution-flow/_index.md:156]: We don't need this script any more.

File: docs/content/contributing/terraform/terraform-contribution-flow/_index.md
Comment (matt-FFFFFF) [inline docs/content/contributing/terraform/terraform-contribution-flow/_index.md:187]: Will use avm script now

File: docs/content/contributing/terraform/terraform-contribution-flow/_index.md
Comment (matt-FFFFFF) [inline docs/content/contributing/terraform/terraform-contribution-flow/_index.md:216]: ./avm pre-commit

File: docs/content/contributing/terraform/terraform-contribution-flow/_index.md
Comment (matt-FFFFFF) [inline docs/content/contributing/terraform/terraform-contribution-flow/_index.md:227]: I would rather we did not run grept manually, we will manage centrally

File: docs/content/contributing/terraform/terraform-contribution-flow/_index.md
Comment (matt-FFFFFF) [inline docs/content/contributing/terraform/terraform-contribution-flow/_index.md:296]: This is incorrect, I will make a PR to change - see #522

**Review by matt-FFFFFF — CHANGES_REQUESTED** [review]
Some changes..

We should allow contributors to work from a branch in the main repo as they have write access. Using a fork is more complex and unnecessary.

**Comment by segraef** [comment]
> Some changes..
> 
> We should allow contributors to work from a branch in the main repo as they have write access. Using a fork is more complex and unnecessary.

As outside collaborator then, fine with that. Another option was to have dev branches in the source repo and letting contributors raise PRs against it, once their fork e2e test are green we merge their fork branch into dev and then raise an obligatory PR with e2e which should go green as well. But yeah granting them access as outside collaborator is a simpler solution.

File: docs/content/contributing/terraform/composition.md
Comment (segraef) [reply docs/content/contributing/terraform/composition.md:155]: So not repeating the specs or completely removing the whole `### Casing` section since it's mentioned in the specs already?

File: docs/content/contributing/terraform/prerequisites.md
Comment (segraef) [reply docs/content/contributing/terraform/prerequisites.md:62]: What to be defined as prerequisites depends on what we recommend for testing/developing a module. I think there is an open space between using the container image `azterraform` (with all tools bootstrapped) and a contributor's local machine which has all predefined tools installed.

I'd like to discuss this point in our planned call so I can properly define the documentation around it. Marking this as `TBD`.

File: docs/content/contributing/terraform/prerequisites.md
Comment (segraef) [reply docs/content/contributing/terraform/prerequisites.md:71]: Belongs to the same conversation as above mentioned, hence marking this as `TBD`.

File: docs/content/contributing/terraform/prerequisites.md
Comment (segraef) [reply docs/content/contributing/terraform/prerequisites.md:74]: Belongs to the same conversation as above mentioned, hence marking this as `TBD`.

File: docs/content/contributing/terraform/terraform-contribution-flow/_index.md
Comment (segraef) [reply docs/content/contributing/terraform/terraform-contribution-flow/_index.md:35]: Only if a user is member of the Azure org which only can be FTEs. Except we follow the suggestion to grant outside collaborators (non-Azure-org members) `Write` on the repository.

File: docs/content/contributing/terraform/terraform-contribution-flow/_index.md
Comment (segraef) [reply docs/content/contributing/terraform/terraform-contribution-flow/_index.md:105]: This stands in relation tho the above mentioned comment about outside collaborators (non-Azure-org). Once we found consensus happy to adjust the order and adjust the wording. Marking this as `TBD`.

File: docs/content/contributing/terraform/terraform-contribution-flow/_index.md
Comment (segraef) [reply docs/content/contributing/terraform/terraform-contribution-flow/_index.md:147]: Marking this as `TBD` in our call.

File: docs/content/contributing/terraform/terraform-contribution-flow/_index.md
Comment (segraef) [reply docs/content/contributing/terraform/terraform-contribution-flow/_index.md:156]: Happy to adjust. Marking this as `TBD` in our call.

File: docs/content/contributing/terraform/terraform-contribution-flow/_index.md
Comment (segraef) [reply docs/content/contributing/terraform/terraform-contribution-flow/_index.md:296]: Removed.

File: docs/content/contributing/terraform/terraform-contribution-flow/_index.md
Comment (segraef) [reply docs/content/contributing/terraform/terraform-contribution-flow/_index.md:187]: Marking this as `TBD` in our call.

File: docs/content/contributing/terraform/terraform-contribution-flow/_index.md
Comment (segraef) [reply docs/content/contributing/terraform/terraform-contribution-flow/_index.md:216]: Marking this as `TBD` in our call.

File: docs/content/contributing/terraform/terraform-contribution-flow/_index.md
Comment (segraef) [reply docs/content/contributing/terraform/terraform-contribution-flow/_index.md:227]: Marking this as `TBD` in our call.

File: docs/content/contributing/terraform/terraform-contribution-flow/owner-contribution-flow.md
Comment (segraef) [reply docs/content/contributing/terraform/terraform-contribution-flow/owner-contribution-flow.md:65]: Marking this as `TBD` in our call.

File: docs/content/contributing/terraform/terraform-contribution-flow/owner-contribution-flow.md
Comment (segraef) [reply docs/content/contributing/terraform/terraform-contribution-flow/owner-contribution-flow.md:67]: Marking this as `TBD` in our call.

File: docs/content/contributing/terraform/terraform-contribution-flow/owner-contribution-flow.md
Comment (segraef) [reply docs/content/contributing/terraform/terraform-contribution-flow/owner-contribution-flow.md:104]: Marking this as `TBD` in our call.

File: docs/content/contributing/terraform/prerequisites.md
Comment (segraef) [reply docs/content/contributing/terraform/prerequisites.md:71]: Removed.

File: docs/content/contributing/terraform/terraform-contribution-flow/owner-contribution-flow.md
Comment (segraef) [reply docs/content/contributing/terraform/terraform-contribution-flow/owner-contribution-flow.md:65]: Removed.

File: docs/content/contributing/terraform/terraform-contribution-flow/owner-contribution-flow.md
Comment (segraef) [reply docs/content/contributing/terraform/terraform-contribution-flow/owner-contribution-flow.md:104]: Removed.

File: docs/content/contributing/terraform/prerequisites.md
Comment (segraef) [reply docs/content/contributing/terraform/prerequisites.md:61]: Removed.

File: docs/content/contributing/terraform/terraform-contribution-flow/_index.md
Comment (segraef) [reply docs/content/contributing/terraform/terraform-contribution-flow/_index.md]: moved to owner flow

File: docs/content/contributing/terraform/terraform-contribution-flow/_index.md
Comment (segraef) [reply docs/content/contributing/terraform/terraform-contribution-flow/_index.md]: Adjusted.

File: docs/content/contributing/terraform/terraform-contribution-flow/owner-contribution-flow.md
Comment (segraef) [reply docs/content/contributing/terraform/terraform-contribution-flow/owner-contribution-flow.md]: Removed.

File: docs/content/contributing/terraform/terraform-contribution-flow/owner-contribution-flow.md
Comment (segraef) [reply docs/content/contributing/terraform/terraform-contribution-flow/owner-contribution-flow.md]: Removed.

File: docs/content/contributing/terraform/terraform-contribution-flow/owner-contribution-flow.md
Comment (segraef) [reply docs/content/contributing/terraform/terraform-contribution-flow/owner-contribution-flow.md]: Removed.

**Comment by segraef** [comment]
> > Some changes..
> > We should allow contributors to work from a branch in the main repo as they have write access. Using a fork is more complex and unnecessary.
> 
> As outside collaborator then, fine with that although I thought the secret access will be fixed the need of secret access eliminated with the transition to self-hosted runners? Another option was to have dev branches in the source repo and letting contributors raise PRs against it, once their fork e2e test are green we merge their fork branch into dev and then raise an obligatory PR with e2e which should go green as well. But yeah granting them access as outside collaborator is a simpler solution.
> 
> Regarding your comments about the doco, I'm going to use our Thursday session to run us through the details and incorporate all your comments, thanks.

All questions were clarified.

File: docs/content/contributing/terraform/composition.md
Comment (matt-FFFFFF) [reply docs/content/contributing/terraform/composition.md:254]: same

**Comment by segraef** [comment]
> Some changes..
> 
> We should allow contributors to work from a branch in the main repo as they have write access. Using a fork is more complex and unnecessary.

Added as per our conversation 3 weeks ago see https://github.com/segraef/Azure-Verified-Modules/blob/cd02c4c73fc42b89420ed6e407b5cabe6538c6d8/docs/content/contributing/terraform/terraform-contribution-flow/_index.md?plain=1#L110

File: docs/content/contributing/terraform/terraform-contribution-flow/_index.md
Comment (matt-FFFFFF) [inline docs/content/contributing/terraform/terraform-contribution-flow/_index.md:163]: For contributors working on fork, we don't need UMIs. They don't support client secrets needed to pass creds into container as shown below.

File: docs/content/contributing/terraform/terraform-contribution-flow/_index.md
Comment (matt-FFFFFF) [inline docs/content/contributing/terraform/terraform-contribution-flow/_index.md:158]: We don't need UMIs as they are not supported for client secrets

**Review by matt-FFFFFF — APPROVED** [review]

**Review by matt-FFFFFF — CHANGES_REQUESTED** [review]
Just one change on UMIs for contributors no longer bring requried

File: docs/content/contributing/terraform/terraform-contribution-flow/_index.md
Comment (segraef) [reply docs/content/contributing/terraform/terraform-contribution-flow/_index.md:158]: leftover commment, removed

**Comment by segraef** [comment]
> Just one change on UMIs for contributors no longer bring requried

leftover comments, removed

**Review by matt-FFFFFF — APPROVED** [review]

**Pull request merged**
